File: test-server/turtle_svg.py
"""turtle_svg.py

Python port of ``src/lib/utils/turtle-svg.ts``.

Renders a sequence of turtle-style drawing commands to an SVG document string.
The server uses this to turn one of the ``src/lib/samples/*.json`` command lists
into a renderable vectorization image served from the ``/resource`` route.

Command semantics (identical to the TS original)
------------------------------------------------
The turtle starts at (0, 0) with a heading of 0 degrees (pointing along +x).
Coordinates are screen-style: +y is down, positive angles rotate clockwise.

  - spin:  rotate in place by ``degrees`` (never draws).
  - line:  move forward ``distance`` along the current heading; draws only when
           ``penDown`` is true.
  - arc:   travel along a circular arc of ``radius`` sweeping ``degrees``; the
           heading changes by ``degrees``. Arcs always draw.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Literal, Sequence, TypedDict, Union


# --- Command types ---------------------------------------------------------


class SpinCommand(TypedDict):
    kind: Literal["spin"]
    degrees: float


class LineCommand(TypedDict):
    kind: Literal["line"]
    distance: float
    penDown: bool


class ArcCommand(TypedDict):
    kind: Literal["arc"]
    radius: float
    degrees: float


DrawCommand = Union[SpinCommand, LineCommand, ArcCommand]


@dataclass(frozen=True)
class RenderOptions:
    """Mirrors the TS ``RenderOptions`` interface (only the fields the server
    exercises are given non-default values)."""

    stroke: str = "#1a1a1a"
    stroke_width: float = 2
    padding: float = 20
    background: str | None = None
    start_heading: float = 0
    flip_y: bool = False
    width: float | None = None
    height: float | None = None


DEG2RAD = math.pi / 180

# Max angular span of a single SVG ``A`` segment. Splitting below 180 keeps the
# large-arc flag unambiguous and handles full 360 circles.
MAX_ARC_SEGMENT_DEGREES = 90


def _fmt(n: float) -> str:
    """Compact but precise-enough number formatting for path data."""
    return str(round(n, 3) + 0.0)  # ``+ 0.0`` collapses ``-0.0`` to ``0.0``


@dataclass
class _Bounds:
    min_x: float
    min_y: float
    max_x: float
    max_y: float


@dataclass
class _Point:
    x: float
    y: float


def _trace_path(commands: Sequence[DrawCommand], start_heading: float) -> tuple[str, _Bounds | None]:
    """Walk the command list once, producing SVG path data and the bounding box
    of everything that was drawn."""
    x = 0.0
    y = 0.0
    heading = float(start_heading)  # degrees

    d = ""
    pen_at: _Point | None = None  # last point emitted into the path, if any

    min_x = math.inf
    min_y = math.inf
    max_x = -math.inf
    max_y = -math.inf

    def include(px: float, py: float) -> None:
        nonlocal min_x, min_y, max_x, max_y
        if px < min_x:
            min_x = px
        if py < min_y:
            min_y = py
        if px > max_x:
            max_x = px
        if py > max_y:
            max_y = py

    def move_to(px: float, py: float) -> None:
        nonlocal d, pen_at
        d += f"M {_fmt(px)} {_fmt(py)} "
        pen_at = _Point(px, py)
        include(px, py)

    def ensure_pen_at(px: float, py: float) -> None:
        # Start a new subpath unless the pen is already exactly here.
        if pen_at is None or pen_at.x != px or pen_at.y != py:
            move_to(px, py)

    for cmd in commands:
        kind = cmd["kind"]
        if kind == "spin":
            heading += cmd["degrees"]

        elif kind == "line":
            rad = heading * DEG2RAD
            nx = x + cmd["distance"] * math.cos(rad)
            ny = y + cmd["distance"] * math.sin(rad)
            if cmd["penDown"]:
                ensure_pen_at(x, y)
                d += f"L {_fmt(nx)} {_fmt(ny)} "
                pen_at = _Point(nx, ny)
                include(nx, ny)
            x = nx
            y = ny

        elif kind == "arc":
            radius = cmd["radius"]
            degrees = cmd["degrees"]
            if radius <= 0 or degrees == 0:
                heading += degrees
                continue

            # The arc's center sits perpendicular to the heading, on the side
            # the turtle is turning toward.
            side = math.copysign(1, degrees)  # +1 or -1
            center_angle = (heading + side * 90) * DEG2RAD
            cx = x + radius * math.cos(center_angle)
            cy = y + radius * math.sin(center_angle)

            # Angle from center to the turtle's current position.
            phi = heading - side * 90

            ensure_pen_at(x, y)

            # Emit the arc in <=90 chunks so full circles and reflex arcs render
            # correctly with large-arc-flag = 0.
            sweep_flag = 1 if degrees > 0 else 0
            chunks = max(1, math.ceil(abs(degrees) / MAX_ARC_SEGMENT_DEGREES))
            step = degrees / chunks

            for i in range(1, chunks + 1):
                phi_end = (phi + step * i) * DEG2RAD
                ex = cx + radius * math.cos(phi_end)
                ey = cy + radius * math.sin(phi_end)
                d += f"A {_fmt(radius)} {_fmt(radius)} 0 0 {sweep_flag} {_fmt(ex)} {_fmt(ey)} "

                # Sample the chunk for the bounding box (arc extrema can lie
                # between endpoints).
                phi_start = (phi + step * (i - 1)) * DEG2RAD
                samples = 8
                for s in range(samples + 1):
                    a = phi_start + ((phi_end - phi_start) * s) / samples
                    include(cx + radius * math.cos(a), cy + radius * math.sin(a))

            phi_final = (phi + degrees) * DEG2RAD
            x = cx + radius * math.cos(phi_final)
            y = cy + radius * math.sin(phi_final)
            pen_at = _Point(x, y)
            heading += degrees

    bounds = None if min_x == math.inf else _Bounds(min_x, min_y, max_x, max_y)
    return d.strip(), bounds


def render_to_svg(commands: Sequence[DrawCommand], options: RenderOptions | None = None) -> str:
    """Render the commands to a complete standalone SVG document string."""
    opts = options or RenderOptions()
    d, bounds = _trace_path(commands, opts.start_heading)

    # Fall back to a small empty canvas if nothing was drawn.
    b = bounds or _Bounds(0, 0, 100, 100)
    vb_x = b.min_x - opts.padding
    vb_y = b.min_y - opts.padding
    vb_w = b.max_x - b.min_x + opts.padding * 2
    vb_h = b.max_y - b.min_y + opts.padding * 2

    aspect = vb_w / vb_h
    out_w = opts.width if opts.width is not None else (opts.height * aspect if opts.height is not None else None)
    out_h = opts.height if opts.height is not None else (opts.width / aspect if opts.width is not None else None)

    size_attrs = (f' width="{_fmt(out_w)}"' if out_w is not None else "") + (
        f' height="{_fmt(out_h)}"' if out_h is not None else ""
    )

    bg_rect = (
        f'<rect x="{_fmt(vb_x)}" y="{_fmt(vb_y)}" width="{_fmt(vb_w)}" '
        f'height="{_fmt(vb_h)}" fill="{opts.background}"/>'
        if opts.background
        else ""
    )

    transform = f' transform="translate(0 {_fmt(vb_y * 2 + vb_h)}) scale(1 -1)"' if opts.flip_y else ""

    return (
        '<svg xmlns="http://www.w3.org/2000/svg"'
        f' viewBox="{_fmt(vb_x)} {_fmt(vb_y)} {_fmt(vb_w)} {_fmt(vb_h)}"{size_attrs}>'
        + bg_rect
        + f'<path d="{d}" fill="none" stroke="{opts.stroke}"'
        + f' stroke-width="{opts.stroke_width}" stroke-linecap="round"'
        + f' stroke-linejoin="round"{transform}/>'
        + "</svg>"
    )

File: src/lib/utils/turtle-svg.ts
/**
 * turtle-svg.ts
 *
 * Renders a sequence of turtle-style drawing commands to an SVG string,
 * and provides helpers to encode that SVG as a data URL suitable for
 * assigning to an <img src="...">.
 *
 * Command semantics
 * -----------------
 * The turtle starts at (0, 0) with a heading of 0° (pointing along +x).
 * Coordinates are screen-style: +y is down, and positive angles rotate
 * clockwise on screen.
 *
 *  - spin:  rotate in place by `degrees` (no movement, never draws).
 *  - line:  move forward `distance` along the current heading.
 *           Draws only when `penDown` is true.
 *  - arc:   travel along a circular arc of `radius`, sweeping `degrees`.
 *           Positive degrees curve toward the turtle's "positive angle"
 *           side, negative degrees curve the other way. The heading
 *           changes by `degrees`. Arcs always draw (the source format
 *           carries no pen flag for them).
 *
 * Usage
 * -----
 *   import { commandsToImgSrc } from "./turtle-svg";
 *   const img = document.createElement("img");
 *   img.src = commandsToImgSrc(commands, { stroke: "#222", strokeWidth: 2 });
 */

export interface SpinCommand {
  kind: "spin";
  degrees: number;
}

export interface LineCommand {
  kind: "line";
  distance: number;
  penDown: boolean;
}

export interface ArcCommand {
  kind: "arc";
  radius: number;
  degrees: number;
}

export type DrawCommand = SpinCommand | LineCommand | ArcCommand;

export interface RenderOptions {
  /** Stroke color for the drawing. Default: "#1a1a1a". */
  stroke?: string;
  /** Stroke width in drawing units. Default: 2. */
  strokeWidth?: number;
  /** Padding around the drawing's bounding box, in drawing units. Default: 20. */
  padding?: number;
  /** Optional background fill. Default: none (transparent). */
  background?: string;
  /** Initial heading in degrees. Default: 0 (pointing along +x). */
  startHeading?: number;
  /**
   * Mirror the drawing vertically (use if your source data assumed a
   * math-style +y-up coordinate system). Default: false.
   */
  flipY?: boolean;
  /** Fixed output size, e.g. { width: 512 }. Height is derived from the
   *  drawing's aspect ratio when omitted. By default the SVG has no
   *  width/height and scales to its container via the viewBox. */
  width?: number;
  height?: number;
}

interface Point {
  x: number;
  y: number;
}

const DEG2RAD = Math.PI / 180;

/** Max angular span of a single SVG `A` segment. Splitting below 180°
 *  keeps the large-arc flag unambiguous and handles full 360° circles. */
const MAX_ARC_SEGMENT_DEGREES = 90;

function fmt(n: number): string {
  // Compact but precise-enough number formatting for path data.
  return Number(n.toFixed(3)).toString();
}

/**
 * Walks the command list once, producing SVG path data and tracking the
 * bounding box of everything that was drawn.
 */
function tracePath(
  commands: readonly DrawCommand[],
  startHeading: number,
): {
  d: string;
  bounds: { minX: number; minY: number; maxX: number; maxY: number } | null;
} {
  let x = 0;
  let y = 0;
  let heading = startHeading; // degrees

  let d = "";
  let penAt: Point | null = null; // last point emitted into the path, if any

  let minX = Infinity;
  let minY = Infinity;
  let maxX = Infinity * -1;
  let maxY = -Infinity;

  const include = (px: number, py: number) => {
    if (px < minX) minX = px;
    if (py < minY) minY = py;
    if (px > maxX) maxX = px;
    if (py > maxY) maxY = py;
  };

  const moveTo = (px: number, py: number) => {
    d += `M ${fmt(px)} ${fmt(py)} `;
    penAt = { x: px, y: py };
    include(px, py);
  };

  const ensurePenAt = (px: number, py: number) => {
    // Start a new subpath unless the pen is already exactly here.
    if (!penAt || penAt.x !== px || penAt.y !== py) moveTo(px, py);
  };

  for (const cmd of commands) {
    switch (cmd.kind) {
      case "spin": {
        heading += cmd.degrees;
        break;
      }

      case "line": {
        const rad = heading * DEG2RAD;
        const nx = x + cmd.distance * Math.cos(rad);
        const ny = y + cmd.distance * Math.sin(rad);
        if (cmd.penDown) {
          ensurePenAt(x, y);
          d += `L ${fmt(nx)} ${fmt(ny)} `;
          penAt = { x: nx, y: ny };
          include(nx, ny);
        }
        x = nx;
        y = ny;
        break;
      }

      case "arc": {
        const { radius, degrees } = cmd;
        if (radius <= 0 || degrees === 0) {
          heading += degrees;
          break;
        }

        // The arc's center sits perpendicular to the heading, on the side
        // the turtle is turning toward.
        const side = Math.sign(degrees); // +1 or -1
        const centerAngle = (heading + side * 90) * DEG2RAD;
        const cx = x + radius * Math.cos(centerAngle);
        const cy = y + radius * Math.sin(centerAngle);

        // Angle from center to the turtle's current position.
        let phi = heading - side * 90;

        ensurePenAt(x, y);

        // Emit the arc in <=90° chunks so full circles and reflex arcs
        // render correctly with large-arc-flag = 0.
        const sweepFlag = degrees > 0 ? 1 : 0;
        const chunks = Math.max(
          1,
          Math.ceil(Math.abs(degrees) / MAX_ARC_SEGMENT_DEGREES),
        );
        const step = degrees / chunks;

        for (let i = 1; i <= chunks; i++) {
          const phiEnd = (phi + step * i) * DEG2RAD;
          const ex = cx + radius * Math.cos(phiEnd);
          const ey = cy + radius * Math.sin(phiEnd);
          d += `A ${fmt(radius)} ${fmt(radius)} 0 0 ${sweepFlag} ${fmt(ex)} ${fmt(ey)} `;

          // Sample the chunk for the bounding box (arc extrema can lie
          // between endpoints).
          const phiStart = (phi + step * (i - 1)) * DEG2RAD;
          const samples = 8;
          for (let s = 0; s <= samples; s++) {
            const a = phiStart + ((phiEnd - phiStart) * s) / samples;
            include(cx + radius * Math.cos(a), cy + radius * Math.sin(a));
          }
        }

        const phiFinal = (phi + degrees) * DEG2RAD;
        x = cx + radius * Math.cos(phiFinal);
        y = cy + radius * Math.sin(phiFinal);
        penAt = { x, y };
        heading += degrees;
        break;
      }
    }
  }

  const bounds = minX === Infinity ? null : { minX, minY, maxX, maxY };
  return { d: d.trim(), bounds };
}

/**
 * Renders the commands to a complete standalone SVG document string.
 */
export function renderToSvg(
  commands: readonly DrawCommand[],
  options: RenderOptions = {},
): string {
  const {
    stroke = "#1a1a1a",
    strokeWidth = 2,
    padding = 20,
    background,
    startHeading = 0,
    flipY = false,
    width,
    height,
  } = options;

  const { d, bounds } = tracePath(commands, startHeading);

  // Fall back to a small empty canvas if nothing was drawn.
  const b = bounds ?? { minX: 0, minY: 0, maxX: 100, maxY: 100 };
  const vbX = b.minX - padding;
  const vbY = b.minY - padding;
  const vbW = b.maxX - b.minX + padding * 2;
  const vbH = b.maxY - b.minY + padding * 2;

  const aspect = vbW / vbH;
  const outW = width ?? (height !== undefined ? height * aspect : undefined);
  const outH = height ?? (width !== undefined ? width / aspect : undefined);

  const sizeAttrs =
    (outW !== undefined ? ` width="${fmt(outW)}"` : "") +
    (outH !== undefined ? ` height="${fmt(outH)}"` : "");

  const bgRect = background
    ? `<rect x="${fmt(vbX)}" y="${fmt(vbY)}" width="${fmt(vbW)}" height="${fmt(
        vbH,
      )}" fill="${background}"/>`
    : "";

  const transform = flipY
    ? ` transform="translate(0 ${fmt(vbY * 2 + vbH)}) scale(1 -1)"`
    : "";

  return (
    `<svg xmlns="http://www.w3.org/2000/svg"` +
    ` viewBox="${fmt(vbX)} ${fmt(vbY)} ${fmt(vbW)} ${fmt(vbH)}"${sizeAttrs}>` +
    bgRect +
    `<path d="${d}" fill="none" stroke="${stroke}"` +
    ` stroke-width="${strokeWidth}" stroke-linecap="round"` +
    ` stroke-linejoin="round"${transform}/>` +
    `</svg>`
  );
}

/**
 * Encodes an SVG document string as a data URL usable in <img src>.
 * Uses URI-encoding (more compact than base64 for SVG text).
 */
export const svgToDataUrl = (svg: string) =>
  `data:image/svg+xml;charset=utf-8,${encodeURIComponent(svg)}`;

/**
 * One-step convenience: commands -> data URL for an <img> tag.
 *
 *   document.querySelector("img")!.src = commandsToImgSrc(commands);
 */
export const commandsToImgSrc = (
  commands: readonly DrawCommand[],
  options?: RenderOptions,
) => svgToDataUrl(renderToSvg(commands, options));
